### ISSUE TITLE
gulp seems to hang with async task

### ISSUE THREAD

**Issue by ethul** [gulpjs/gulp#167]
hi,

I might be missing something here, but let's say I define a task as follows.

``` javascript
gulp.task('aaa', function(cb){
  http.get('http://www.google.com/index.html', function(res){
    gutil.log('in result');
    cb();
  });
});
```

If I run this task from default and then invoke gulp from the command line I get

``` bash
prompt$ gulp
[gulp] Running 'default'...
[gulp] Running 'aaa'...
[gulp] Finished 'default' in 4.9 ms
[gulp] in result
[gulp] Finished 'aaa' in 134 ms
```

But the process hangs there and I don't get back to the prompt.

However, If redefine the task as follows, and run it

``` javascript
gulp.task('aaa', function(cb){
  setTimeout(cb, 2000);
});
```

``` bash
prompt$ gulp
[gulp] Running 'default'...
[gulp] Running 'aaa'...
[gulp] Finished 'default' in 833 μs
[gulp] Finished 'aaa' in 2 s
prompt$
```

This works as expected and the process terminates.

Not sure what I am missing, but any pointers would be great.

**Comment by jfairley** [comment]
Having the same problem when I run a [gulp-istanbul](https://github.com/SBoudrias/gulp-istanbul) test. @nidheeshdas, your solution works great for me. Would love to understand what I'm not seeing however.

```javascript
gulp.task('test', ['env:test', 'jshint'], function () {
    var deferred = Q.defer();
    gulp
        .src(['app/**/*.js', '!app/**/*_test.js', 'config/**/*.js'])
        .pipe(istanbul()) // Covering files
        .pipe(istanbul.hookRequire()) // Force `require` to return covered files
        .on('finish', function () {
            return gulp.src(['app/**/*_test.js', 'spec/**/*.js'])
                .pipe(mocha())
                .pipe(istanbul.writeReports({
                    reporters: ['lcov', 'cobertura', 'text']
                })) // Creating the reports after tests run
                .pipe(istanbul.enforceThresholds({
                    thresholds: {
                        global: {
                            branches: 75,
                            functions: 80,
                            lines: 80,
                            statements: 80
                        }
                    }
                }))
                .on('error', deferred.reject)
                .on('end', deferred.resolve);
        });
    return deferred.promise;
});
```

**Comment by ivanixgames** [comment]
Hello,
I am seeing hanging behavior specifically when using the gulp-shell plugin in when it is used to call gulp recursively to process different subdirectories.  It hangs on in my Mac OSX but runs without hanging on my Ubuntu box.

**Comment by karlroberts** [comment]
nidheeshdas solution worked for me, I don't understand how or why as I've just got into Gulp.
any further comment to explain it would be nice

my watch task is below:-

    /* Notes:
       - gulp/tasks/browserify.js handles js recompiling with watchify
       - gulp/tasks/browserSync.js watches and reloads compiled files
    */

    var gulp     = require('gulp');
    var config   = require('../config');

    var isWatching = false;

    gulp.task('watch', ['watchify','browserSync'], function() {
      isWatching = true;
      gulp.watch(config.sass.src,   ['sass']);
      gulp.watch(config.images.src, ['images']);
      gulp.watch(config.markup.src, ['markup']);
      // Watchify will watch and recompile our JS, so no need to gulp.watch it
    });

    gulp.on('stop', function() {
        if (!isWatching) {
            process.nextTick(function() {
                process.exit(0);
            });
        }
    });

**Comment by kkworden** [comment]
This happens to me as well, Ubuntu 14.10. I need to manually put process.exit(0), otherwise my gulpfile will never finish, even after all my tasks are completed.

**Comment by pavankumarkatakam** [comment]
Same problem here.

**Comment by ynk** [comment]
just got the same with an async task which is kinda long (external build process, ~50s+).
could be related to a timeout maybe ?

**Comment by trevorhreed** [comment]
I periodically experience a hang when running gulp tasks. It's happened during watch task as well as a simple build task. When I hit `Ctrl+c` it seems to suddenly catch up (displaying messages it should have displayed if not for the hang) and then, of course, immediately shuts down. I know it's not just the messages that are being delayed because I've checked the files that are building. It's as if, when I hit `Ctrl+c`, I stop some hanging process that the rest of the gulp task is waiting on, then the rest of the gulp task runs, quickly, before the program shuts down.

**Comment by SunTzu92** [comment]
Não sei se isso ajuda ainda mas, no meu caso estava chamando uma task em forma de loop, ocorrendo esse evento:

`[17:42:48] Using gulpfile ~/www/consorcio/gulpfile.js`

e minha task estava assim:

`gulp.task('ex1',  ['ex2'], () => {
    return del(config.ex);
});

gulp.task('ex2',  ['ex1'], () => {
    return del(config.ex2);
});`

o que fazia entrar em loop na hora do build default

e no meu watch:

`gulp.task('sync', () => {
    gulp.watch(config.ex, ['ex1']);
    gulp.watch(config.ex2, ['ex2']);
});`

resolvir apos remover a dependencia de ambos, e o gulp voltou a funcionar normamente.
espero ter ajudado de alguma forma.

**Comment by dewwwald** [comment]
@nidheeshdas nice solution.

You can monkey patch watch for cleaner code.
```
const watch = gulp.watch;
gulp.watch = function (...args) {
  watching = true;
  return watch(...args);
};

gulp.on('stop', function () {
  if (!watching) {
    process.nextTick(function () {
      process.exit(0);
    });
  }
});
```
